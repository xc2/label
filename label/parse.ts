import { InvalidLabel } from "./exceptions";
import type { AbsoluteLabel, Label } from "./label";

function determineTarget(target: string, packageLastPart: string): string {
  if (target) {
    return target;
  } else {
    if (packageLastPart === "...") {
      return target;
    } else {
      // //foo
      return packageLastPart;
    }
  }
}

/**
 * Parse a label string into a `Label` object
 * @public
 * @param label - the label string to parse
 * @throws {@link InvalidLabel} if the label is invalid
 * @returns The parsed `Label` object
 */
export function parse(label: string | Label): Label {
  if (typeof label !== "string") {
    return label;
  }
  const [match, _scope, _repoName, _packagePath, _targetName] =
    label.match(
      //12222222211111  3333333333333333333    444444
      /^(([^\/]+)?\/\/)?([^:.]*(?:\.\.\.)?)(?::([^:]+))?$/
    ) || [];
  if (!match) {
    throw new InvalidLabel(label);
  }
  const packageLastPart = _packagePath.split("/").pop() || "";
  const includeSubPackages = packageLastPart === "...";
  const packagePath = includeSubPackages ? _packagePath.slice(0, -4) : _packagePath;

  const target = determineTarget(_targetName, packageLastPart);

  return {
    scope: _scope ? _repoName || "" : false,
    package: packagePath,
    includeSubPackages,
    target,
  } as Label;
}

/**
 * Check if a label is with a scope
 * @param label - the label to check
 * @throws {@link InvalidLabel} if the label is not an absolute label
 * @public
 */
export function validateAbsolute(label: string | Label): AbsoluteLabel {
  const l = parse(label);
  if (l.scope !== false) {
    return l as AbsoluteLabel;
  }
  throw new InvalidLabel(label, "must be an absolute label");
}
