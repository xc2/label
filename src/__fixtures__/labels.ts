export const labels = [
  "//foo/bar:wiz",
  "//foo/bar",
  "//foo/bar:all",
  "//foo/...",
  "//foo/...:all",
  "//foo/...:*",
  "//foo/...:all-targets",
  "//...",
  "//:all",
  ":foo",
  "bar:wiz",
  "bar/wiz",
  "bar:all",
  ":all",
  "...:all",
  "...",
  "bar/...:all",
];

export const invalids = ["//foo/.:a", "//foo:a:b"];
