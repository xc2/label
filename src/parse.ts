import { InvalidLabel } from "./exceptions.js";
import type { AbsoluteLabel, Label } from "./label.js";

function determineTarget(target: string, packageLastPart: string): string {
  if (target) {
    return target;
  } else {
    if (packageLastPart === "...") {
      return target;
    } else {
      // //foo
      return packageLastPart;
    }
  }
}

export function parse(label: string | Label): Label {
  if (typeof label !== "string") {
    return label;
  }
  const [match, _scope, _repoName, _packagePath, _targetName] =
    label.match(
      //12222222211111  3333333    4444
      /^(([^\/]+)?\/\/)?([^:]*)(?::(.+))?/
    ) || [];
  if (!match) {
    throw new InvalidLabel(label);
  }
  const packageLastPart = _packagePath.split("/").pop() || "";
  const includeSubPackages = packageLastPart === "...";
  const packagePath = includeSubPackages ? _packagePath.slice(0, -4) : _packagePath;

  const target = determineTarget(_targetName, packageLastPart);

  return {
    scope: _scope ? _repoName || "" : false,
    package: packagePath,
    includeSubPackages,
    target,
  } as Label;
}

export function parseAbsolute(label: string | Label): AbsoluteLabel {
  const l = parse(label);
  if (l.scope !== false) {
    return l as AbsoluteLabel;
  }
  throw new InvalidLabel(label, "must be an absolute label");
}
