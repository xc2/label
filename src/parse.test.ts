import { expect, test } from "bun:test";
import { labels } from "./__fixtures__/labels";
import { parse } from "./parse.js";

test.each(labels)("parse (%p)", (label) => {
  expect(parse(label)).toMatchSnapshot();
});
