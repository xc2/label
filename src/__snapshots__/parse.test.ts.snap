// Bun Snapshot v1, https://goo.gl/fbAQLP


exports[`parse ("//foo/bar:wiz") 1`] = `
{
  "includeSubPackages": false,
  "package": "foo/bar",
  "scope": "",
  "target": "wiz",
}
`;

exports[`parse ("//foo/bar") 1`] = `
{
  "includeSubPackages": false,
  "package": "foo/bar",
  "scope": "",
  "target": "bar",
}
`;

exports[`parse ("//foo/bar:all") 1`] = `
{
  "includeSubPackages": false,
  "package": "foo/bar",
  "scope": "",
  "target": Symbol(All Rules Targets),
}
`;

exports[`parse ("//foo/...") 1`] = `
{
  "includeSubPackages": true,
  "package": "foo",
  "scope": "",
  "target": Symbol(All Rules Targets),
}
`;

exports[`parse ("//foo/...:all") 1`] = `
{
  "includeSubPackages": true,
  "package": "foo",
  "scope": "",
  "target": Symbol(All Rules Targets),
}
`;

exports[`parse ("//foo/...:*") 1`] = `
{
  "includeSubPackages": true,
  "package": "foo",
  "scope": "",
  "target": Symbol(All Targets (Rules and Files)),
}
`;

exports[`parse ("//foo/...:all-targets") 1`] = `
{
  "includeSubPackages": true,
  "package": "foo",
  "scope": "",
  "target": Symbol(All Targets (Rules and Files)),
}
`;

exports[`parse ("//...") 1`] = `
{
  "includeSubPackages": true,
  "package": "",
  "scope": "",
  "target": Symbol(All Rules Targets),
}
`;

exports[`parse ("//:all") 1`] = `
{
  "includeSubPackages": false,
  "package": "",
  "scope": "",
  "target": Symbol(All Rules Targets),
}
`;

exports[`parse (":foo") 1`] = `
{
  "includeSubPackages": false,
  "package": "",
  "scope": Symbol(Relative so unknown),
  "target": "foo",
}
`;

exports[`parse ("bar:wiz") 1`] = `
{
  "includeSubPackages": false,
  "package": "bar",
  "scope": Symbol(Relative so unknown),
  "target": "wiz",
}
`;

exports[`parse ("bar/wiz") 1`] = `
{
  "includeSubPackages": false,
  "package": "bar/wiz",
  "scope": Symbol(Relative so unknown),
  "target": "wiz",
}
`;

exports[`parse ("bar:all") 1`] = `
{
  "includeSubPackages": false,
  "package": "bar",
  "scope": Symbol(Relative so unknown),
  "target": Symbol(All Rules Targets),
}
`;

exports[`parse (":all") 1`] = `
{
  "includeSubPackages": false,
  "package": "",
  "scope": Symbol(Relative so unknown),
  "target": Symbol(All Rules Targets),
}
`;

exports[`parse ("...:all") 1`] = `
{
  "includeSubPackages": true,
  "package": "",
  "scope": Symbol(Relative so unknown),
  "target": Symbol(All Rules Targets),
}
`;

exports[`parse ("...") 1`] = `
{
  "includeSubPackages": true,
  "package": "",
  "scope": Symbol(Relative so unknown),
  "target": Symbol(All Rules Targets),
}
`;

exports[`parse ("bar/...:all") 1`] = `
{
  "includeSubPackages": true,
  "package": "bar",
  "scope": Symbol(Relative so unknown),
  "target": Symbol(All Rules Targets),
}
`;
